src/js/views/OperationsLog/useValidation.jsx

```diff
@@ -0,0 +1,49 @@
+import { useState } from 'react'
+import { useTranslation } from 'react-i18next'
+import { validate as jsonschemaValidate } from 'jsonschema'
+import { jsonSchema } from '../../schema/OperationsLog'
+
+export function useValidation() {
+  const [errors, setErrors] = useState({})
+  const { t } = useTranslation()
+
+  function errorMessage(field) {
+    switch (field) {
+      case 'change_type':
+        return t('operationsLog.validation.changeTypeError')
+      case 'environment':
+        return t('operationsLog.validation.environmentError')
+      case 'recorded_at':
+        return t('operationsLog.validation.recordedAtError')
+      case 'completed_at':
+        return t('operationsLog.validation.completedAtError')
+      case 'description':
+        return t('operationsLog.validation.descriptionError')
+      case 'project':
+        return t('operationsLog.validation.projectError')
+      case 'version':
+        return t('operationsLog.validation.versionError')
+      case 'ticket_slug':
+        return t('operationsLog.validation.ticketSlugError')
+      case 'link':
+        return t('operationsLog.validation.linkError')
+      case 'notes':
+        return t('operationsLog.validation.notesError')
+      default:
+        return t('error.title')
+    }
+  }
```

Comment:
I'm all for making this more generic. And the call to `setErrors()` is perfectly fine.

The tricky part of not separating out `setErrors()` from `validate()` is that now the logic can't be inlined in `onSubmit()`, `onSubmit()` would need to instead trigger validation and then stop. A separate `useEffect` would need to listen for validation changes, then make the POST/PATCH request if everything is ok.

I actually think that's cleaner, because it doesn't expect the client to "just know" that it needs to create the errors, and then also set them when it's done. But it does require some extra bookkeeping - the client needs to know if we just did a validation so we can either submit or not. We can leverage the `saving` state variable for this. Although, ideally, I think I'd like all this logic in the form. It could be a state machine that knows if it's in the `edit`, `post-validate` or `submit` states. But I'll start with just this and let's see how it looks.